conn.go

```diff
@@ -383,6 +405,16 @@ func (c *Conn) shutdown() func() error {
 	defer c.mtx.Unlock()
 	runtime.LockOSThread()
 	defer runtime.UnlockOSThread()
+	timed_out := false
+	time.AfterFunc(300*time.Millisecond, func() {
```

Comment:
Where does this value come from?

Reply:
@marten-seemann It's a local variable meant to prevent an infinite loop.

Reply:
First, I don't understand where an infinite loop would come from (I don't understand the whole `handshake_started` / `handshake_finished` logic in this PR.

Second, starting a timer here is definitely not the solution. 300 ms seem to be an arbitrary value, and it will just cause problems later on. Furthermore, this code is racy anyway (you're writing to and reading from `timed_out` from separate go routines).

Reply:
@marten-seemann The point of the timeout loop is that it prevents a connection being closed while a handshake is in progress, which occurs in the C portion of the OpenSSL library. I consistently run into that error in my libp2p-tls PR. I can remove this from this PR for now, but it will still be necessary I think. Also, I can't test how well this PR helps with libp2p integration until this is merged.

Assuming that Go booleans are equivalent to C booleans once compiled, they are atomic (modified and read in 1 instruction) and thus don't actually have any race conditions.